src/blk/kernel/io_uring.cc

```diff
@@ -3,7 +3,7 @@
 
 #include "io_uring.h"
 
-#if defined(HAVE_LIBURING) && defined(__x86_64__)
+#if defined(HAVE_LIBURING) && (defined(__x86_64__) || defined(__aarch64__))
```

Comment:
i think it's safe to drop `defined(__x86_64__)`, as `HAVE_LIBURING` ensures us the availability of liburing. as i think liburing is based on io_uring which is a feature of linux kernel, and hence is available on more archs, like armv7 and i586/i686. see also https://packages.gentoo.org/packages/sys-libs/liburing and https://packages.gentoo.org/packages/sys-libs/liburing. 

@rouming what do you think?

Reply:
It depends on the uring lib and support of syscalls for different archs (since there is no glibc support each syscall should be explicitly defined for each arch). I do not know the current uring lib code, but I'm pretty sure if arch is not supported explicit enosupport error should be returned (at least I clearly picture that in my mind). So let's give it a try :)

Reply:
`
root@ubuntu:~/gitee/fio# dpkg -l | grep liburing
ii  liburing-dev:arm64                    0.7-1                              arm64        userspace library for using io_uring
ii  liburing1:arm64                       0.7-1                              arm64        userspace library for using io_uring
root@ubuntu:~/gitee/fio# ./fio -ioengine=io_uring -numjobs=1 -direct=1 -size=512g -iodepth=1024 -runtime=300 -rw=randwrite -ba=1m -bs=1m -filename=/dev/nvme0n1 -hipri -name=iscsi_io_uring_1024_randwrite_1m -group_reporting
iscsi_io_uring_1024_randwrite_1m: (g=0): rw=randwrite, bs=(R) 1024KiB-1024KiB, (W) 1024KiB-1024KiB, (T) 1024KiB-1024KiB, ioengine=io_uring, iodepth=1024
fio-3.23-48-ga0b72-dirty
Starting 1 process
Jobs: 1 (f=0)
yujiang o->hipri:  1

yujiang syscall(__NR_io_uring_setup, depth, &p) ret:  6
Jobs: 1 (f=0): [f(1)][100.0%][w=1976MiB/s][w=1975 IOPS][eta 00m:00s]
iscsi_io_uring_1024_randwrite_1m: (groupid=0, jobs=1): err= 0: pid=466236: Fri Nov 20 08:45:50 2020
  write: IOPS=2414, BW=2414MiB/s (2531MB/s)(512GiB/217178msec); 0 zone resets
    slat (usec): min=98, max=4000, avg=267.80, stdev=58.34
    clat (msec): min=246, max=1036, avg=423.69, stdev=34.53
     lat (msec): min=246, max=1036, avg=423.95, stdev=34.53
    clat percentiles (msec):
     |  1.00th=[  363],  5.00th=[  376], 10.00th=[  388], 20.00th=[  401],
     | 30.00th=[  414], 40.00th=[  418], 50.00th=[  422], 60.00th=[  430],
     | 70.00th=[  435], 80.00th=[  443], 90.00th=[  456], 95.00th=[  464],
     | 99.00th=[  550], 99.50th=[  584], 99.90th=[  718], 99.95th=[  785],
     | 99.99th=[  877]
   bw (  MiB/s): min= 1654, max= 3538, per=100.00%, avg=2416.93, stdev=166.26, samples=433
   iops        : min= 1654, max= 3538, avg=2416.93, stdev=166.26, samples=433
  lat (msec)   : 250=0.05%, 500=98.08%, 750=1.79%, 1000=0.08%, 2000=0.01%
  cpu          : usr=31.54%, sys=68.45%, ctx=201, majf=0, minf=5618112
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.1%
     issued rwts: total=0,524288,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=1024

Run status group 0 (all jobs):
  WRITE: bw=2414MiB/s (2531MB/s), 2414MiB/s-2414MiB/s (2531MB/s-2531MB/s), io=512GiB (550GB), run=217178-217178msec

Disk stats (read/write):
  nvme0n1: ios=46/524020, merge=0/0, ticks=8/221157926, in_queue=220111220, util=99.40%
root@ubuntu:~/gitee/fio# lscpu
Architecture:                    aarch64
CPU op-mode(s):                  64-bit
Byte Order:                      Little Endian
CPU(s):                          64
On-line CPU(s) list:             0-63
Thread(s) per core:              1
Core(s) per socket:              4
Socket(s):                       16
NUMA node(s):                    8
Vendor ID:                       0x70
Model:                           2
Stepping:                        0x1
BogoMIPS:                        100.00
NUMA node0 CPU(s):               0-7
NUMA node1 CPU(s):               8-15
NUMA node2 CPU(s):               16-23
NUMA node3 CPU(s):               24-31
NUMA node4 CPU(s):               32-39
NUMA node5 CPU(s):               40-47
NUMA node6 CPU(s):               48-55
NUMA node7 CPU(s):               56-63
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Mitigation; __user pointer sanitization
Vulnerability Spectre v2:        Vulnerable
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fp asimd evtstrm crc32 cpuid
`
Yes, I tried aarch64 and it can run iouring.

Reply:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/standalone/ceph-helpers.sh:666: run_osd:  ceph-osd -i 0 --fsid=439c7f56-0687-4095-9d4c-0f25887b907f --auth-supported=none --mon-host=127.0.0.1:54621 --osd-failsafe-full-ratio=.99 --osd-journal-size=100 --osd-scrub-load-threshold=2000 --osd-data=td/smoke/0 --osd-journal=td/smoke/0/journal --chdir= --run-dir=td/smoke '--admin-socket=/tmp/ceph-asok.12183/$cluster-$name.asok' --debug-osd=20 --debug-ms=1 --debug-monc=20 '--log-file=td/smoke/$name.log' '--pid-file=td/smoke/$name.pid' --osd-max-object-name-len=460 --osd-max-object-namespace-len=64 '--enable-experimental-unrecoverable-data-corrupting-features=*' --mkfs --key AQDFY7dfXUWADxAA7OzOIu8L5dhbQ/EA4qfKNg== --osd-uuid 0f73a9e8-e13e-4bab-a425-5f2bbf01e616
2020-11-20T06:35:50.002+0000 7f00fe319140 -1 WARNING: all dangerous and experimental features are enabled.
2020-11-20T06:35:50.030+0000 7f00fe319140 -1 WARNING: all dangerous and experimental features are enabled.
2020-11-20T06:35:50.046+0000 7f00fe319140 -1 WARNING: all dangerous and experimental features are enabled.
2020-11-20T06:35:50.046+0000 7f00fe319140 -1 bluestore(td/smoke/0/block) _read_bdev_label failed to open td/smoke/0/block: (2) No such file or directory
2020-11-20T06:35:50.046+0000 7f00fe319140 -1 bluestore(td/smoke/0/block) _read_bdev_label failed to open td/smoke/0/block: (2) No such file or directory
2020-11-20T06:35:50.046+0000 7f00fe319140 -1 bluestore(td/smoke/0/block) _read_bdev_label failed to open td/smoke/0/block: (2) No such file or directory
2020-11-20T06:35:50.050+0000 7f00fe319140 -1 bluestore(td/smoke/0) _read_fsid unparsable uuid 
```

Reply:
@lnsyyj could you consider the suggestion at https://github.com/ceph/ceph/pull/38204#discussion_r527502788 ?

Reply:
@tchaikov 
Judging from the Available Versions in https://packages.gentoo.org/packages/sys-libs/liburing, currently liburing->kernel iouring only supports amd64, x86, arm, arm64 instruction sets. I think from the current kernel iouring progress, it is recommended not to remove x86 from the ceph layer, but to add a pre-compiled macro for every CPU instruction set tested.

Reply:
> `HAVE_LIBURING` ensures us the availability of liburing

i think this is a valid reason why we don't need to check the archs.

Reply:
Oh! I understand what you mean. Agree to remove `defined(__x86_64__)`.